exonum_client/api.py

```diff
@@ -0,0 +1,552 @@
+import requests
+
+from .protobuf_loader import ProtoFile
+
+# pylint: disable=C0103
+logger = getLogger(__name__)
+
+
+class Api:
+    """Api class provides basic REST functionality."""
+
+    # constants
+    RUST_RUNTIME_ID = 0
+
+    def __init__(self, hostname: str, port: int, ssl: bool = False):
+        """
+        Constructor of Api.
+
+        Parameters
+        ----------
+        hostname: str
+            Examples: '127.0.0.1', 'www.some_node.com'.
+        port: int
+            API port of an Exonum node.
+        ssl: bool
+            If True, HTTPS protocol is used for communication, otherwise HTTP is used.
+        """
+        self.schema = "https" if ssl else "http"
+        self.hostname = hostname
+        self.port = port
+        # Example of a formatted prefix: "https://127.0.0.1:8000"
+        self.endpoint_prefix = "{}://{}:{}/api".format(self.schema, hostname, port)
+
+    @staticmethod
+    def get(url: str, params: Optional[Dict[Any, Any]] = None) -> requests.Response:
+        """Internal wrapper over requests.get"""
+        return requests.get(url, params=params)
+
+    @staticmethod
+    def post(url: str, data: str, headers: Dict[str, str]) -> requests.Response:
+        """Internal wrapper over requests.post"""
+        return requests.post(url, data=data, headers=headers)
+
+
+class PublicApi(Api):
+    """PublicApi class provides methods to interact with the public API of an Exonum node."""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self._tx_url = self.endpoint_prefix + "/explorer/v1/transactions"
+        self._rust_runtime_url = self.endpoint_prefix + "/runtimes/rust/{}"
+        self._block_url = self.endpoint_prefix + "/explorer/v1/block"
+        self._blocks_url = self.endpoint_prefix + "/explorer/v1/blocks"
+        self._system_url = self.endpoint_prefix + "/system/v1/{}"
+
+    # Implementation of ProtobufProviderInterface:
+    def _get_proto_sources(self, params: Optional[Dict[str, str]] = None) -> List[ProtoFile]:
+        proto_sources_endpoint = self._rust_runtime_url.format("proto-sources")
+        response = self.get(proto_sources_endpoint, params=params)
+        if response.status_code != 200 or "application/json" not in response.headers["content-type"]:
+            logger.critical(
+                "Unsuccessfully attempted to retrieve Protobuf sources.\n" "Status code: %s,\n" "body:\n%s",
+                response.status_code,
+                json.dumps(response.json(), indent=2),
+            )
+            raise RuntimeError("Unsuccessfully attempted to retrieve Protobuf sources: {!r}".format(response.content))
+        logger.debug("Protobuf sources retrieved successfully.")
+
+        proto_files = [
+            ProtoFile(name=proto_file["name"], content=proto_file["content"]) for proto_file in response.json()
+        ]
+
+        return proto_files
+
+    def get_main_proto_sources(self) -> List[ProtoFile]:
```

Comment:
Those methods are implementation of `ProtobufProviderInterface`. I think you can create a separate class, like `ProtobufApi` which will derive from both `ProtobufProviderInterface` and `Api` for this purpose, since those methods aren't really intended for use by clint clients (pun not intended).